core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,415 @@
+    unless the collection has already been completed or has been created/edited
+    by the user. It is also removed from the incomplete list (if present).
+
+    Args:
+        user_id: str. The id of the user who completed the collection.
+        collection_id: str. The id of the completed collection.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    subscribed_collection_ids = (
+        subscription_services.get_collection_ids_subscribed_to(user_id))
+
+    activities_completed = get_activities_completed_from_model(
+        activities_completed_model)
+
+    if (collection_id not in subscribed_collection_ids and
+            collection_id not in activities_completed.completed_collection_ids):
+        # Remove the collection from the in progress list (if present) as it is
+        # now completed.
+        remove_collection_from_incomplete_list(user_id, collection_id)
+        activities_completed.add_collection_id(collection_id)
+        save_activities_completed(activities_completed)
+
+
+def mark_exploration_as_incomplete(
+        user_id, exploration_id, state_name, exploration_version):
+    """Adds the exploration id to the incomplete list of the user unless the
+    exploration has been already completed or has been created/edited by the
+    user. If the exploration is already present in the incomplete list, just the
+    details associated with it are updated.
+
+    Args:
+        user_id: str. The id of the user who partially completed the
+            exploration.
+        timestamp: datetime.datetime. The time at which the user left the
+            exploraion.
+        exploration_id: str. The id of the partially completed exploration.
+        state_name: str. The name of the state at which the user left the
+            exploration.
+        exploration_version: str. The version of the exploration played by the
+            learner.
+    """
+    incomplete_activities_model = (
+        user_models.IncompleteActivitiesModel.get(
+            user_id, strict=False))
+    if not incomplete_activities_model:
+        incomplete_activities_model = (
+            user_models.IncompleteActivitiesModel(id=user_id))
+
+    completed_exploration_ids = get_all_completed_exp_ids(user_id)
+
+    subscribed_exploration_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    incomplete_activities = get_incomplete_activities_from_model(
+        incomplete_activities_model)
+
+    if (exploration_id not in completed_exploration_ids and
+            exploration_id not in subscribed_exploration_ids):
+
+        if exploration_id not in incomplete_activities.incomplete_exploration_ids: # pylint: disable=line-too-long
```

Comment:
You can put parens around this and break after "in" to avoid the line-too-long pragma.